### DISCUSSION TITLE
SSH with 2FA

### DISCUSSION THREAD

**Discussion by Aariq** [mschubert/clustermq#280] [Q&A]
Recently my HPC turned on 2-factor authentication for SSH access.  I had this working well for a while—running `Q()` from RStudio brought up a dialog box to enter an option for how to authenticate, then push went to the app on my phone.  I'm not sure what changed, but now when I run `Q()`, the console hangs.  Sometimes, **after** I try to terminate the stuck R process the 2FA dialog box pops up. Is there any chance this is something on the `clustermq` side, or is this likely to be an issue with the 2FA?

**Comment by Aariq** [comment]
I followed the debugging guide and got stuck with this step:

```
system("ssh -R 54709:localhost:6687 <user@host> R --vanilla")
```

It hangs, and when I interrupt the R session, the dialog box for 2FA comes up.  If I add `wait = FALSE` it appears to work though

**Reply by mschubert** [reply]
Thanks, that likely also explains behavior in https://github.com/mschubert/clustermq/issues/206. I'm not sure what the best solution is yet, because if we set `wait=FALSE` a forward failure will not be caught.

There was no change on the `clustermq` side (CRAN) since 0.8.95 (July 2020).

**Comment by Aariq** [comment]
Ok, so my workflow with `targets` and `tar_make_clustermq()` is working again and I honestly don't know what changed or if something was actually broken.  I wanted to comment though because the debugging step, `system("ssh -R 54709:localhost:6687 <user@host> R --vanilla")`, led me astray.  Running this from R running in a terminal session works, but I need to enter an option for 2FA, like so:

```
> system("ssh -R 54709:localhost:6687 <user@host> R --vanilla")
(<user@host>) Duo two-factor login for <user>

Enter a passcode or select one of the following options:

 1. Duo Push to XXX-XXX-5555
 2. Phone call to XXX-XXX-5555

Passcode or option (1-2): 

```

I enter an option, then authenticate on my phone, and I'm in.  But if I run this R code from R Studio either in the console pane, by clicking "Source" in the editor pane, or by running the line from the editor with command-enter, it hangs, then only after I terminate the R session does a dialog box pop up with the text above.  If I save the code as a .R script and use  "Source as a Local Job..." the dialog box pops up and the code runs.  So the hanging I experienced with this debugging step wasn't actually related to `clustermq` working or not.  It seems like it's maybe actually bug in RStudio.

**Reply by Aariq** [reply]
Oh, and `tar_make_clustermq()` works fine from the console, editor, or "source as a local job...", by the way.

**Reply by mschubert** [reply]
Thanks, this is helpful! Would it work for you if I change the recommendation to test `system("ssh -R 54709:localhost:6687 <user@host> R --version")`?

**Reply by Aariq** [reply]
This seems to have the same behavior.  I'm guessing any shell command that prompts the user for something before it finishes does this in RStudio.  I don't know enough about shell scripting to get a minimal reprex though.
